### PR TITLE
Fix runtime stack trace computation

### LINKED ISSUE #8880: assert points to the wrong line
### What version of Bun is running?

1.0.26+c75e768a6

### What platform is your computer?

Darwin 23.2.0 arm64 arm

### What steps can reproduce the bug?

This file:

```ts
import assert from 'assert'

function calc() {
  return 1
}

assert.strictEqual(1, 2)
```

### What is the expected behavior?

assertion error points to the right line

### What do you see instead?

Screenshot 2024-02-13 at 04 50 19

### Additional information

_No response_

### LINKED ISSUE #9302: Source maps and error locations are broken
The following is an overview over bugs within Bun's source map implementation. This bug is the root cause behind the following:

- bun run
  - Error locations are not pointing at the right character, or line.
    - Error locations can be off by hundreds of lines in the worst case.
  - Bun's VCSode extension (and the work JetBrains is putting to get debugging to work in their products) is unable to break on many lines.
- bun build
  - Browser debuggers are unable to break on many lines
  - Sourcemap analysis tools are unable to open some of Bun's generated source maps.

I've given some efforts to fix these, but I've realized none of this is a quick fix, and will need a much larger audit over every bit of sourcemap code.

## Incorrect Sourcemaps
Sourcemaps are generated incorrectly. A full pass over sourcemap generation in js_parser.zig or bundle_v2.zig should be done to figure out what is going wrong.

Figure A *(there are two spaces after the `-`)*
```ts
;(hello -  world);
```
Figure B
```ts
hello-
world
```

esbuild produces different sourcemaps despite identical JS output.
This is significant because remapping sources does not map the same, some maps are off, some do not map at all.

`demo.sh`
```ts
bun build '/Users/dave/hello/min.js' --target=browser --outdir=bun --sourcemap=external
bunx esbuild '/Users/dave/hello/min.js' --platform=browser --outdir=esbuild --sourcemap=external
bun dump_maps.js
diff --color bun.map esbuild.map
```
`dump_maps.js`
```ts
import { SourceMapConsumer} from 'source-map';
import { readFileSync, writeFileSync } from 'fs';

async function dump(from, to, off = 0) {
    const data = JSON.parse(readFileSync(from, 'utf8'));
    await SourceMapConsumer.with(data, null, async(map) => {
        let out = '';
        map.eachMapping((m) => {
            out += `${m.source} ${m.originalLine}:${m.originalColumn} -> ${m.generatedLine + off}:${m.generatedColumn}\n`;
        });
        await writeFileSync(to, out);
    });
}

await dump('bun/min.js.map', 'bun.map', -1);
await dump('esbuild/min.js.map', 'esbuild.map', );
```

This is more than just adding brace close locations (the initial assumption), but a more fundamental issue in the sourcemaps.

## JSC vs V8 Stack Differences
This seems to only be visible when accessing error.stack. No idea why; More investigation is needed.

While not a issue in source maps, having 1:1 stack traces with `error.stack` is a goal of Bun, especially since we modified `error.stack` to print a v8 style stack trace.

In JSC/Safari, the column number of function calls are at the index before the `(`, where in v8/Node/Deno, this is located at the start of the identifier.

```ts
  hello()
//^    ^
//v8   jsc
```

This difference is visible in Bun's `error.stack` but only there, doing `console.log` (our formatter instead of error.stack) on the error works fine. I'm confused why this is only an issue in one part, considering JSC in no place reports the v8 style error column numbers.

## 1.0.15 Regression

Potentially caused by source mapping bugs, this is a regression in 1.0.15, when I was first made aware of how broken the sourcemaps are. I am not marking this issue as a whole as a regression, as I haven't found a single build of bun where the sourcemap generation was correct.

Figure C
```ts
(() => {
    function functionWithAName() {
        Error.prepareStackTrace = (e, s) => {
          console.log( s[0].getLineNumber());
          console.log( s[0].getFunctionName());
          console.log( s[1].getLineNumber());
        };
        const e = new Error();
        console.log(e.stack);
    }

    functionWithAName();
})();

// $ bun c.js # before, and node
// 8
// functionWithAName
// 12
// undefined

// $ bun c.js # after
// 8
// functionWithAName
// 10
// undefined
```

## Incorrect error.stack without prepareStackTrace

Figure D
```ts
(() => {
    function functionWithAName() {
        Error.prepareStackTrace = (e, s) => {
          console.log( s[0].getLineNumber());
          console.log( s[0].getFunctionName());
          console.log( s[1].getLineNumber());
        };
        const e = new Error();
        console.log(e.stack);
    }

    functionWithAName();
})();
```
Node.js
```
Error
    at functionWithAName (/d.js:8:19)
    at /d.js:12:5
    at Object.<anonymous> (/d.js:13:3)
```
Bun 1.0.14
```
Error: 
    at functionWithAName (/d:8:27)
    at <anonymous> (/d.js:12:3)
    at module code (/d.js:7:3)
```
Bun 1.0.15
```
Error
    at functionWithAName (/d.js:8:27)
    at <anonymous> (/d.js:6:20)
    at module code (/d.js:7:3)
```
In all cases Bun is remapping to neither what JSC reports nor what v8 would report.
The issue is invisible in bun if you `console.log` the error

```
3 |         // Error.prepareStackTrace = (e, s) => {
4 |         //   console.log( s[0].getLineNumber());
5 |         //   console.log( s[0].getFunctionName());
6 |         //   console.log( s[1].getLineNumber());
7 |         // };
8 |         const e = new Error();
                      ^
error: Error
      at functionWithAName (/d:8:19)
      at /d.js:12:5
      at /d.js:12:23
```

Actually, that last mapping is incorrect, but that one is likely the closing brace loc issue, as it is incorrect remapping the IIFE call `)()` using the `functionWithAName()` call invocation mapping data.

### PR DESCRIPTION
**Pull request by paperclover**
Fixes #9302
Fixes #8880

After the previous adventure of source-mapping related issues, it seems the big issue blocking sourcemaps from working is simply our code reading line+column numbers from JavaScriptCore.

The original code
```diff
-   /* Get the "raw" position info.
-    * Note that we're using m_codeBlock->unlinkedCodeBlock()->expressionRangeForBytecodeOffset
-    * rather than m_codeBlock->expressionRangeForBytecodeOffset in order get the "raw" offsets and
-    * avoid the CodeBlock's expressionRangeForBytecodeOffset modifications to the line and column
-    * numbers, (we don't need the column number from it, and we'll calculate the line "fixes"
-    * ourselves). */
-   ExpressionInfo::Entry info = m_codeBlock->unlinkedCodeBlock()->expressionInfoForBytecodeIndex(bytecodeOffset);
```

Is an alright idea, but the code that handles fixing this raw info does not work, with cases as simple as:

```
// @bun
var l = Object.create; (() => { throw new Error() })();
```

Instead we will simply use `expressionInfoForBytecodeIndex` on the linked code block to retrive all of the computed offsets.

```diff
+auto expr = m_codeBlock->expressionInfoForBytecodeIndex(bytecodeOffset);
```

And to prevent zero-based vs one-based, I am using `OrdinalNumber` practically everywhere to ensure these numbers are interpretted correctly.

Draft PR because there are surely test failures which will need to be corrected with proper location info, but I am noticing some potential mistakes in regards to oneBasedInt and zeroBasedInt

### CODE REVIEW

**Comment by paperdave** [comment]
TODO: this regresses this behavior
image

**Comment by paperdave** [comment]
This seems to move the column number from the `new` keyword to the `Error` function name.

image

**Comment by paperdave** [comment]
> This seems to move the column number from the `new` keyword to the `Error` function name.

This one is actually insane.
image
Look at this sourcemap, if the `Error(` is one source mapping token, then it means a divot pointing at the ( will remap to point at the `E` in Error, which is what we observe.

The old logic was either
- bugged in just the right way to work
- we had an off-by-one token issue

I'm going to work on getting the sourcemap columns on function calls to just *work*. Matching v8 is so close i can taste it.

File: src/bun.js/bindings/headers-handwritten.h [hunk 1]
```diff
@@ -138,14 +139,9 @@ const ZigStackFrameCode ZigStackFrameCodeWasm = 5;
...
 typedef struct ZigStackFramePosition {
-    int32_t source_offset;
-    int32_t line;
-    int32_t line_start;
-    int32_t line_stop;
-    int32_t column_start;
-    int32_t column_stop;
-    int32_t expression_start;
-    int32_t expression_stop;
+    WTF::OrdinalNumber line;
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/headers-handwritten.h]: Is it safe to use `WTF::OrdinalNumber` here? It won't be initialized as an C++ `WTF::OrdinalNumber` in Zig. I think this should be stored as an int32_t and then a method could be added to convert it to a WTF::OrdinalNumber.

File: src/bun.js/bindings/headers-handwritten.h [hunk 2]
```diff
@@ -138,14 +139,9 @@ const ZigStackFrameCode ZigStackFrameCodeWasm = 5;
...
+    WTF::OrdinalNumber column;
+    int byte_position;
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/headers-handwritten.h]: Let's be more specific so there's no ambiguity:
```suggestion
    int32_t byte_position;
```

File: src/bun.js/bindings/headers-handwritten.h [hunk 3]
```diff
@@ -158,7 +154,7 @@ typedef struct ZigStackFrame {
 
 typedef struct ZigStackTrace {
     BunString* source_lines_ptr;
-    int32_t* source_lines_numbers;
+    OrdinalNumber* source_lines_numbers;
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/headers-handwritten.h:167]: We can't share C++ classes in Zig
```suggestion
    int32_t* source_lines_numbers;
```

**Review by Jarred-Sumner — COMMENTED** [review]
A couple nitpicky comments but this is a good improvement and once the tests pass & the comments are addressed, we should merge

File: src/bun.js/bindings/headers-handwritten.h (hunk 1 above)
Comment (paperclover) [reply src/bun.js/bindings/headers-handwritten.h]: yeah lets do this

File: src/bun.js/bindings/ErrorStackFrame.cpp [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    if (pos.byte_position - amount < 0) {
+        pos.line_zero_based = 0;
+        pos.column_zero_based = 0;
+        pos.byte_position = 0;
+        return;
+    }
+
+    pos.column_zero_based = pos.column_zero_based - amount;
+    if (pos.column_zero_based < 0) {
+        auto source = code->source().provider()->source();
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/ErrorStackFrame.cpp:25]:
```suggestion
        const auto *provider = code->source().provider();
        if (UNLIKELY(!provider)) {
           return;
        }
        const auto& source = provider->source();
```

**Pull request merged**